src/relax/transform/utils.cc

```diff
@@ -22,17 +22,43 @@
 namespace tvm {
 namespace relax {
 
+bool IsScalarTensor(const StructInfo& sinfo) {
+  if (!sinfo->IsInstance<TensorStructInfoNode>()) {
+    return false;
+  }
+  TensorStructInfo tensor_sinfo = Downcast<TensorStructInfo>(sinfo);
+  if (!tensor_sinfo->shape.defined() || !tensor_sinfo->shape->IsInstance<ShapeExprNode>()) {
+    return false;
+  }
+  return tensor_sinfo->shape.as<ShapeExprNode>()->values.size() == 0;
+}
+
+bool IsScalarTensor(const Expr& expr) { return IsScalarTensor(GetStructInfo(expr)); }
+
 bool IsNestedTensor(const StructInfo& sinfo) {
   if (sinfo->IsInstance<TensorStructInfoNode>()) {
     return true;
   } else if (const auto* tuple_sinfo = sinfo.as<TupleStructInfoNode>()) {
     return !std::any_of(tuple_sinfo->fields.begin(), tuple_sinfo->fields.end(),
-                        [&](const StructInfo& field) { return !IsNestedTensor(field); });
+                        [](const StructInfo& field) { return !IsNestedTensor(field); });
   }
   return false;
 }
 
 bool IsNestedTensor(const Expr& expr) { return IsNestedTensor(GetStructInfo(expr)); }
 
+bool IsNestedFloatTensor(const StructInfo& sinfo) {
+  if (auto* tensor_sinfo = sinfo.as<TensorStructInfoNode>()) {
+    return tensor_sinfo->dtype.is_float() || tensor_sinfo->dtype.is_float16() ||
+           tensor_sinfo->dtype.is_bfloat16();
+  } else if (auto* tuple_sinfo = sinfo.as<TupleStructInfoNode>()) {
+    return !std::any_of(tuple_sinfo->fields.begin(), tuple_sinfo->fields.end(),
+                        [](const StructInfo& field) { return !IsNestedFloatTensor(field); });
+  }
+  return false;
+}
+
+bool IsNestedFloatTensor(const Expr& expr) { return IsNestedFloatTensor(GetStructInfo(expr)); }
```

Comment:
I suppose we can turn this into a generic one called `IsNestedTensorConditioned`, which takes an Expr/StructInfo as input, together with a function that returns a boolean value to indicate “if the leaf tensor meets the condition.”

The impl would be
```c++
template <typename FType>
bool IsNestedTensorConditioned(const StructInfo& sinfo, FType f_condition) {
  if (auto* tensor_sinfo = sinfo.as<TensorStructInfoNode>()) {
    return f_condition(GetRef<TensorStructInfo>(tensor_sinfo));
  } else if (auto* tuple_sinfo = sinfo.as<TupleStructInfoNode>()) {
    return !std::any_of(
        tuple_sinfo->fields.begin(), tuple_sinfo->fields.end(),
        [](const StructInfo& field) { return !IsNestedTensorConditioned(field, f_condition); });
  }
  return false;
}
```

And the plain IsNestedTensor could be implemented by IsNestedTensorConditioned, with `f_condition` returns true always.